extensions/vertx-redis/runtime/src/main/java/io/quarkus/vertx/redis/runtime/RedisAPIProducer.java

```diff
@@ -0,0 +1,119 @@
+import io.quarkus.vertx.redis.SyncRedisAPI;
+import io.vertx.core.Vertx;
+import io.vertx.redis.client.Redis;
+import io.vertx.redis.client.RedisAPI;
+import io.vertx.redis.client.RedisClientType;
+import io.vertx.redis.client.RedisOptions;
+
+@ApplicationScoped
+class RedisAPIProducer {
+    private static final char AT = '@';
+    private static final char COLON = ':';
+    private static final char SLASH = '/';
+    private static final String REDIS_SCHEME = "redis://";
+    private long timeout = 10;
+
+    private final Redis redisClient;
+
+    private final RedisAPI redisAPI;
+
+    private final SyncRedisAPI syncRedisAPI;
+
+    private final io.vertx.mutiny.redis.client.Redis mutinyRedisClient;
+
+    private final io.vertx.mutiny.redis.client.RedisAPI mutinyRedisAPI;
+
+    public RedisAPIProducer(RedisConfig config, Vertx vertx) {
+        RedisOptions options = new RedisOptions();
+        options.setType(config.clientType);
+
+        if (RedisClientType.STANDALONE == config.clientType) {
+            if (config.hosts.isPresent() && config.hosts.get().size() > 1) {
+                throw new ConfigurationException("Multiple hosts supplied for non clustered configuration");
+            }
+        }
+
+        if (config.hosts.isPresent()) {
+            Set<InetSocketAddress> hosts = config.hosts.get();
+            for (InetSocketAddress host : hosts) {
+                options.addConnectionString(buildRedisUrl(host, config.password, config.database));
+            }
+        } else {
+            InetSocketAddress defaultRedisAddress = new InetSocketAddress("localhost", 6379);
+            options.addConnectionString(buildRedisUrl(defaultRedisAddress, config.password, config.database));
+        }
+
+        if (config.timeout.isPresent()) {
+            timeout = config.timeout.get().getSeconds();
+        }
+
+        redisClient = Redis.createClient(vertx, options);
+        redisAPI = RedisAPI.api(redisClient);
+        mutinyRedisClient = io.vertx.mutiny.redis.client.Redis.newInstance(redisClient);
+        mutinyRedisAPI = io.vertx.mutiny.redis.client.RedisAPI.api(mutinyRedisClient);
+        syncRedisAPI = new SyncRedisAPIImpl(mutinyRedisAPI, timeout);
+    }
+
+    @Produces
+    @Singleton
+    Redis redis() {
+        return redisClient;
+    }
+
+    @Produces
+    @Singleton
+    RedisAPI redisAPI() {
+        return redisAPI;
+    }
+
+    @Produces
+    @Singleton
+    SyncRedisAPI syncRedisAPI() {
+        return syncRedisAPI;
+    }
+
+    @Produces
+    @Singleton
+    io.vertx.mutiny.redis.client.Redis mutinyRedisClient() {
+        return mutinyRedisClient;
+    }
+
+    @Produces
+    @Singleton
+    io.vertx.mutiny.redis.client.RedisAPI mutinyRedisAPI() {
+        return mutinyRedisAPI;
+    }
+
+    @PreDestroy
+    public void close() {
+        this.redis().close();
+    }
+
+    private String buildRedisUrl(InetSocketAddress address, Optional<String> password, int database) {
```

Comment:
Redistribution has now an official iana registered connection string url format. Upstream we have the parser under the impl package but if it makes sense you could refer to it `RedisURI` as the api is stable and shouldn't change.

Reply:
Great. Good to know. I'll look into that.